libtransmission/peer-mgr.h

```diff
@@ -316,17 +344,43 @@ class tr_peer_info
         return ret;
     }
 
-private:
-    [[nodiscard]] constexpr tr_address const& addr() const noexcept
-    {
-        return socket_address_.address();
-    }
+    // ---
 
-    [[nodiscard]] constexpr auto port() const noexcept
+    // only to be used if our peer reported a listening port via ltep handshake
+    void merge_connectable_into_incoming(tr_peer_info const& connectable) noexcept
     {
-        return socket_address_.port();
+        connection_attempted_at_ = std::max(connection_attempted_at_, connectable.connection_attempted_at_);
+        /* connection_changed_at_ is already the latest */
+        /* piece_data_at_ is already the latest */
+
+        /* no need to merge blocklist since it gets updated elsewhere */
+
+        TR_ASSERT(!is_connectable_);
+        if (auto& conn = connectable.is_connectable(); conn && *conn)
+        {
+            set_connectable();
+        }
+        // rationale for not setting "not connectable" even if the peer was flagged as such:
+        // we thought the peer is not connectable, but here it is connecting to us, and reporting that it is listening
+        // so we better try again for ourselves
+
+        /* is_utp_supported_ is already the latest */
+
+        /* from_first_ is already the latest */
+        found_at(connectable.from_best());
+
+        /* num_consecutive_fails_ is already the latest */
+        /* pex flags are already the latest */
+
+        if (connectable.is_banned())
+        {
+            ban();
+        }
+        /* is_connected_ should already be set */
+        /* is_seed_ should already be the latest */
```

Comment:
Note to reviewers: This is probably the most sketchy part of this PR. Would appreciate any feedback or bug reports.